alf/drivers/off_policy_async_driver.py

```diff
@@ -0,0 +1,543 @@
+from alf.drivers.threads import TFQueues, ActorThread, EnvThread, LogThread
+from alf.drivers.threads import get_initial_policy_state
+from alf.drivers.threads import get_initial_time_step
+from alf.drivers.threads import repeat_shape_n
+from alf.drivers.threads import algorithm_step, get_act_dist_param
+from alf.drivers.threads import transform_observation
+from alf.drivers.threads import flatten_once
+
+from alf.algorithms.rl_algorithm import make_training_info
+
+"""
+A TEMPORARY async off-policy driver. To be merged with off_policy_driver.py.
+"""
+
+
+def summarize_time(name, data):
+    with tf.name_scope("Elapsed times (seconds)"):
+        summarize_single_scalars(name=name, data=data)
+
+
+def summarize_single_scalars(name, data):
+    if not tf.executing_eagerly():
+        data.set_shape(())
+    return tf.summary.scalar(name=name, data=data)
+
+
+def transpose2(x, dim1, dim2):
+    perm = list(range(len(x.shape)))
+    perm[dim1] = dim2
+    perm[dim2] = dim1
+    return tf.transpose(x, perm)
+
+
+@six.add_metaclass(abc.ABCMeta)
+class ExperienceReplayer(object):
+    """
+    Base class for implementing experience storing and replay. A subclass
+    should implement two abstract functions: `observe` and `replay`.
+    This class object will be used by OffPolicyAsyncDriver after training
+    data are dequeued from the learning queue.
+    """
+    @abc.abstractmethod
+    def observe(self, exp):
+        """observe an `exp`, potentially storing it to replay buffers"""
+
+    @abc.abstractmethod
+    def replay(self):
+        """Replay experiences from buffers"""
+
+
+@gin.configurable
+class OffPolicyAsyncDriver(object):
+    """
+    A driver that enables async training. The definition of 'async' is that prediction
+    and learning are not intervened like below:
+        pred -> learn -> pred -> learn -> ...
+    Instead they are decoupled:
+        pred -> pred -> pred -> ...
+                   |
+            (sync periodically)
+                   |
+        learn -> learn -> learn -> ...
+    And more importantly, a learner or predictor may only operate on a subset of
+    environments at a time.
+    """
+
+    def __init__(self,
+                 env_f: Callable,
+                 algorithm: OffPolicyAlgorithm,
+                 num_envs=32,
+                 num_act_queues=4,
+                 max_num_iterations=1000000,
+                 max_num_env_steps=100000000,
+                 unroll_length=8,
+                 learn_queue_cap=8,
+                 act_queue_cap=8,
+                 num_updates_per_train_step=1,
+                 mini_batch_length=None,
+                 mini_batch_size=None,
+                 greedy_predict=False,
+                 observation_transformer: Callable = None,
+                 observers=[],
+                 metrics=[],
+                 exp_replayer_class=None,
+                 debug_summaries=False,
+                 summarize_grads_and_vars=False,
+                 training=True,
```

Comment:
what is the use case for training=False?

Reply:
removed. Originally used for unittest of metrics